File: 01-数据结构/01-数组/go/LC 数组中的第K个最大的元素.go
package main

func findKthLargest(nums []int, k int) int {
    return findKth(nums, 0, len(nums)-1, k)
}

func findKth (nums []int, l int, r int, k int) int {
    n := partion(nums,l,r)

    if n == len(nums) - k {
        return nums[n]
    }else if n < len(nums) - k {
        return findKth(nums, n+1, r, k)
    }else {
        return findKth(nums, l, n-1, k)
    }
}

func partion(nums []int, l int, r int) int {
    povt := nums[l]
    left, right := l, r
    for l < r {
        for l < right && nums[l] <= povt {
            l++
        }
        for left < r && nums[r] >= povt {
            r--
        }

        if l < r {
            nums[l], nums[r] = nums[r], nums[l]
            l++
            r--
        }
    }

    nums[left], nums[r] = nums[r], nums[left]
    // fmt.Println(nums,r)
    return r;
}

File: 01-数据结构/01-数组/c++/LC 数组中的第K个最大的元素.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {

        return findKth(nums,0, nums.size()-1, k);

    }

    int findKth(vector<int>& nums, int l, int r, int k) {
        int n = partion(nums,l,r);
        if (n == nums.size() - k) {
            return nums[n];
        }else if (n < nums.size() - k){
            return findKth(nums,n+1,r,k);
        }else {
            return findKth(nums,l,n-1,k);
        }
    }

    int partion(vector<int>& nums, int l, int r) {
        int povt = nums[l];
        int povtIndex = l;
        int left = l;
        int right = r;

        while (l < r) {
            while (l < right && nums[l] <= povt) l++;
            while (left < r && nums[r] >= povt) r--;

            if (l < r) {
                int temp = nums[l];
                nums[l] = nums[r];
                nums[r] = temp;
                l++;
                r--;
            }
        }
        int temp = nums[r];
        nums[r] = nums[povtIndex];
        nums[povtIndex] = temp;
        // for (auto x : nums) {
        //     cout<<x<<" ";
        // }
        // cout<<"-> "<<povt<<endl;
        return r;
    }
};